### ISSUE TITLE
[Metricbeat] Kafka module refactoring

### ISSUE THREAD

**Issue by ChrsMark** [elastic/beats#14852]
## Current situation
Currently Kafka module connects to the local broker and retrieve information only from there. Given the fact of the distributed nature of Kafka, this might be problematic in many cases that we need some ["cluster-wide" metrics](https://github.com/elastic/beats/issues/3608). In addition problems have been reported in cases of topics with [many partitions](https://github.com/elastic/beats/issues/13380).

## Goal of this issue
In https://github.com/elastic/beats/pull/14822 we introduced the use of a "cluster-wide" [client](https://github.com/Shopify/sarama/blob/afedecade3c6d8e99ab6dfeeea7814bf800b90a4/client.go#L15) which connects to a given broker but can retrieve information for the whole cluster through that.

This issue aims to refactor the module so as to:

- [ ] make use of the cluster-wide [client](https://github.com/Shopify/sarama/blob/afedecade3c6d8e99ab6dfeeea7814bf800b90a4/client.go#L15) where possible as described in this [comment](https://github.com/elastic/beats/pull/14822#discussion_r351730994)
- [ ] solve https://github.com/elastic/beats/issues/13380
- [ ] improve `consumer_lag` calculation as described on this [comment](https://github.com/elastic/beats/pull/14822#discussion_r351750303)

@jsoriano @exekias @mtojek feel free to add/comment anything!

**Comment by mtojek** [comment]
@ChrsMark @jsoriano 

The source code looks complex in terms of error handling, retries and waiting for results, so let me rephrase this description to fully understand what should be done/achieved:

1. Metricsets `partition` and `consumergroup` (belonging to `kafka` module) should ALWAYS use `sarama.Client` while perfoming `Fetch()`, to get a reference to the leading broker for a given topic partition.
2. The same broker can be a leader for a different topic partitions. It's fair point to optimize here and combine requests for offsets to several topic partitions.
3. Speaking about `consumergroup`: first fetch partition offsets from leader, then compare it with consumer group status to calculate lag. The consumer group status can be requested from the leading broker for the topic partition?

It looks like a bigger refactoring (incl. tests), hence wondering if isn't it easier to rewrite the code.

**Comment by ChrsMark** [comment]
> @ChrsMark @jsoriano
> 
> The source code looks complex in terms of error handling, retries and waiting for results, so let me rephrase this description to fully understand what should be done/achieved:
> 
> 1. Metricsets `partition` and `consumergroup` (belonging to `kafka` module) should ALWAYS use `sarama.Client` while perfoming `Fetch()`, to get a reference to the leading broker for a given topic partition.

I would say that yes, this is the ideal situation, to make use of the "cluster-wide" client if possible. This will most cover points 1 and 2 from the initial description.

> 2. The same broker can be a leader for a different topic partitions. It's fair point to optimize here and combine requests for offsets to several topic partitions.

I guess that any kind of optimisation is more than welcome :).

> 3. Speaking about `consumergroup`: first fetch partition offsets from leader, then compare it with consumer group status to calculate lag. The consumer group status can be requested from the leading broker for the topic partition?
> 

Maybe https://github.com/Shopify/sarama/blob/master/admin.go#L85 ?

> It looks like a bigger refactoring (incl. tests), hence wondering if isn't it easier to rewrite the code.

I would agree with a rewriting. If I'm not mistaken the key question to answer before this is if we can replace the old client with the "cluster-wide" one. If yes then all the functionalities could be re-written on top of this concept.

**Comment by mtojek** [comment]
@ChrsMark Thank you for answers!

**Comment by jsoriano** [comment]
+1 to rewrite these metricsets if we can simplify the code by using more features of the sarama client and less custom code.

**Comment by mtojek** [comment]
Lessons learnt here:

Ok, so I tried to solve this first and second issue (use cluster-wide client, "broker is not leader"), but apparently grouping requests made the codebase relatively complex and the original issue ("broker is not leader" due to not fresh metadata) might still occur.

After a discussion with Jaime, the way to go might be putting some retries around fetching topic metadata, but this can also make the code not readable.

I will leave this issue open, unless there are no easy steps to solve it (comparing to the total gain)

**Comment by jlind23** [comment]
Backlog grooming: Closing for now.
